learningequality/studio
# Icons, fonts, colors, oh my! Also, more Design System in Studio

## Description

Initially went to use the "Noto Sans" font throughout Studio and to apply font colors per Design System to Studio but along the way ended up making more changes than expected due to how Vuetify handles themes and colors and to avoid going through the entire app to apply styles to every component directly to make use of the proper KDS colors and text styles.

- Vuetify theme now applies the Design System theme tokens over the previously used ones (overrides things like `primary` but keeps other apparently Studio-specific tokens)
- Applies the proper text colors by using the theme-generated CSS variables. This means that all of the tokens defined in https://github.com/learningequality/kolibri-design-system/blob/v0.2.x/lib/styles/colorsDefault.js are available as CSS vars in Studio. For example, the `text` token can be used in CSS on Studio as `--v-text-(base | lighten# | darken#)` the same as other Vuetify theme colors are.
- Implements `KIconButton` by placing it inside of the `IconButton` component (previously wrapped `v-tooltip` and `v-btn` together) and update some token keys uses previously targeting `v-icon` to use the Design System icon token names. This results in a different appearance of the tooltip for icon buttons by using the Design System tooltips. 
- Applied "Noto Sans" across the front end. In some cases, I kept the previous fonts as fall backs. Some places I left alone - like the Markdown Editor and similar areas where `serif`  or symbol fonts were used.
- Misc fixes to fix name-collisions between Vuetify and KDS token names and the styles applied with them by default in Vuetify (particularly the `background-color` on `.link, .text` classes.)
- Adding `KIconButton` ended up being more work than I expected but by the time I realized it was going to be so much more I was kind of pot committed and finished it up. Tokens are renamed, Design System has a PR adding missing icons https://github.com/learningequality/kolibri-design-system/pull/110, and I had to tweak some of the code to handle the fact that we're not putting a `v-btn` in the cards. This resulted in some additional logic around what happens when you click a channel/content/topic card or the buttons within it (copy, info, context, etc). There were resultant style issues that I fixed as well. 

- I also fixed an issue where the badge for a topic in the Clipboard (the purple thing saying how many children there are) was not centered. Just flagging this separately in case someone knows there is a Notion bug-bash issue (I'll give a look later on).

#### Issue Addressed (if applicable)

Fixes https://github.com/learningequality/studio/issues/2264

## Steps to Test

*Icons & Icon Buttons*

- Generally, it is a good idea to just play with every aspect of the app. Settings, channel edit, content edit, exercise editor, 
- Please go through the app everywhere that you know there is an Icon Button of any kind and please let me know if any are weirdly styled, improperly colored, or just don't do the thing they're supposed to do when clicked/hovered/etc. 
- Please verify that the info, copy, star and context menus for ChannelItem work in all cases (not just "My Channels" but also the public library channels (read only channels). Do this for items in the Clipboard as well. Also please do this on the Channel list and in ChannelEdit where content and topics are listed.
- Test this logged in as well as not logged in ("Explore without an account") and ensure all of the buttons work and are where they ought to be.
- Generally keep an eye out for any kind of weird text styles or color inconsistencies.
- Do you see any fonts other than Noto Sans? 


## Implementation Notes (optional)

#### At a high level, how did you implement this?

**Icon Buttons**

I completely replaced the template for `IconButton` and made it a wrapper for `KIconButton` instead. I did this instead of replacing every implementation of `IconButton` in Studio with `KIconButtons` because I want to get thoughts from the team before spending the time changing that many files - also this will give us a chance to see if `KIconButton` can basically just be plugged in wherever we previously used the Studio-custom `IconButton`

The inspiration for this may be hard to explain but I'll try:

- I overrode the Vuetify text color, which should apply also to icons.
- There was no `color` prop for `IconButton` - despite my seeing it be passed somewhere - and even with it properly implemented, it was passed to `v-btn` which does not apply it to the icon properly.
- I want to use the `themeTokens` from KDS for the colors now to avoid having to do it later, so I cannot dynamically override which styles are used depending on what prop value is passed
- `KIconButton` already solves this basic problem by passing the `color` props to children directly to style them directly.

`KIconButton` - however - works exactly how we want it to:

- Tooltip ✓
- Color prop that is applied directly to the icon ✓
- Does everything the `v-btn` did with events ✓

**I add some styles that directly override the Vuetify styles**

I'm not sure if this will have undesired consequences but I did not see any *except* for the one that inspired me to use `KIconButton`.

#### Does this introduce any tech-debt items?

1. Before merging this, I will go through and remove `IconButton` and replace every use of it with `KIconButton`. So - no once I do that.
